### ISSUE TITLE
Add Customer Tax ID

### ISSUE THREAD

**Issue by torvista** [zencart/zencart#3058]
Before starting to use Zen Cart in 2008, I had to add the handling of a customer tax code (create/edit account/orders/invoices).
In Spain an "invoice" is not an invoice (legally) unless it has both supplier and customer tax code present (and ex-vat, vat, total fields).
I think there are several plugins that use a tax code, I don't use any of them.
So, at least I would like a field to be assigned in the customer table as a start point for PRs to reference and plugins to update to.
Or, am I the only one that needs this?
Discuss.

**Comment by DocBenc** [comment]
No you are not alone.. further in Italy (and Portugal too) have now the electronic invoicing (nice) so whenever you generate an invoice it is handled by the central data exchange system (hold by national revenue agency), sent to the customer and then automatically keep in storage for 10 years.
To play this nifty game we need 3 fields: VAT, Tax , Unique ID Code (for electronic invoice).
These are not painless mods as all these data must be fixed in the order table to be "fiscally" corretc. 
The Electronic Invoice will be the standard for EU in the next few years.

**Comment by torvista** [comment]
> To play this nifty game we need 3 fields: VAT, Tax , Unique ID Code (for electronic invoice).

Do you really mean something different to VAT by "Tax" or did you mean to say the base (ex-vat) value?
So are you invoicing from Zen Cart or another app/ERP (as we do)?
In the long term all of this needs to be core, for the reasons you mention, so a start of defining a  customer tx ID field is long overdue, in my opinion.
Possibly also needed is a field to store the validity true/false of the VAT number being validated as intra-community. Since we do not yet do B2B on our store, I have not implemented that yet.

**Comment by DocBenc** [comment]
In short we have 3 tax fields:
1) Tax Code - (Codice Fiscale) assigned to every Italian citizen at birth
2) VAT - the standard EU VAT when you start a business company
3) Unique ID code - (Codice Univoco) for electronic invoicing since 1 Jan 2019

The Tax Code is something italian specific, although other countries have similar code.
The Unique ID code must be collected when customer signin in then transmitted to the ERP. Here in Italy an invoice (fiscally and legally) does not exists without electronic submission through the SDI system (SDI means Exchange Data System) so the uID code is mandatory. From a certain point of view invoicing inside Zencart is totally useless as it has no legal value, it is just an order summary, what we call "courtesy invoice" .

From what i know Italy and Portugal are the first two EU countries to adopt Electronic Invoice and soon this system will be the standard for EU (Germany in 01/2021, Spain is on the route):
https://eur-lex.europa.eu/legal-content/EN/ALL/?uri=celex%3A32006L0112

Electronic invoicing has many repercussions on the management of the tax part, no more paper, no more PDF, no more before-check (all is done by the SDI), no more physical invoices storage, only XML signed invoice files that you can destroy once submitted to the SDI.
You can read more here:
https://www.fieldfisher.com/en/insights/mandatory-electronic-invoicing-in-italy-are-companies-ready

To implement this system i have heavily modified the core system and the core db fields, then implemented a more standardized webservice system with highly reusable code where to make call to defined endpoint and standard Json encoding both for requests and for responses. To be honest i have used these webservice code to do a lot of things also for marketing.

So Zencart certainly needs more native fields to accommodate the near future EU invoicing rules.

**Comment by torvista** [comment]
Excellent information, thanks.
Can you provide a summary of the fields you have needed to add to meet these requirements...you may be one of few that have actually done this in the real world.

**Comment by DocBenc** [comment]
TABLE_ORDERS
billing_vat  :  VAT code
billing_cf  :   TAX code (Codice Fiscale in italian so the shortened _cf)
billing_sdi  :  UId code (_Unique code for SDI)

TABLE_ADDRESS_BOOK
entry_vat
entry_cf
entry_sdi 

Obviously a lot of files alterations are needed:
class orders
pages and modules for login, billing address, customers address book.
I have done all this work via the plugin/observer classes where possible trying to not touch core files, but due to the DB structure of Zencart this is quite difficult to achieve as these mods are functional to all sales.

**Comment by torvista** [comment]
So you use both a personal ID (TAX) and a company ID (VAT) on the same invoice, or these refer to the vendor/issuer of the invoice (you/TAX) and the customer (VAT)?

**Comment by DocBenc** [comment]
For sake of completeness in Italy there was a site "officially" supporting Zencart, the two fields _vat and _cf are historically their modification (although i think it is a porting of a Zencart plugin). 

The sdi fields has been done by myself as some customers needed to be compliant with electronic invoicing from the start. Further the handling of VAT, CF, SDI has been completely reworked (done from scratch) by me as the 3 fields are linked for a Company and must be handled together during registration process.

**Comment by DocBenc** [comment]
> So you use both a personal ID (TAX) and a company ID (VAT) on the same invoice, or these refer to the vendor/issuer of the invoice (you/TAX) and the customer (VAT)?

No TAX id and VAT id refer to customer, they must be present together with SDI id to invoice.

I know that TAX id (Codice fiscale) is something specific to Italy but other countries could use similar code.

**Comment by torvista** [comment]
> No TAX id and VAT id refer to customer, they must be present together with SDI id to invoice.

So an invoice to a **company** as a customer must also include the name and tax id of a physical person in that company!
I've never seen that.
Who would that person be (legally)?

**Comment by DocBenc** [comment]
No there are various types of companies, however they all must have TAX id and VAT id.

For certain types of companies TAX id and VAT id are coincident as they represent a physical tax entity (so however a TAX id must be assigned) but for the current italian laws they must enter both fields separately. These kinds of companies have a physical person representing them but legally are physical tax entities with their own TAX id and VAT id coincident.

So e.g. if you own such a company (called S.r.l very common) you register yourself with your name as sales representative and then ask the invoice with 4 mandatory fields:
1 Company Name
2 Company Vat id
3 Company Tax id (the same value of VAT id)
4 SDI code
These are the only necessary 4 fields to invoice.
I know it is a little bit tricky but these are italian rules. 
I don't ask a customized mods for italy to be included in Zencart but i think that for future Zencart release some fields should be added to accommodate incoming EU rules.

**Comment by torvista** [comment]
>  i think that for future Zencart release some fields should be added to accommodate incoming EU rules.
The more discussion the better, adding a field is no big deal, as long as the intended use is strictly defined.

**Comment by DocBenc** [comment]
Steve maybe this is not the right place but.. news about 1.5.7? Rumors on the release date?

**Comment by torvista** [comment]
https://github.com/zencart/zencart/issues/2973

**Comment by torvista** [comment]
1) Thinking about the VAT field:
is it possible a company may have a tax number but is NOT registered for EU VAT (as has no need for import/export).

2) In the case of an invoice to a private customer (not registered for EU VAT), the TAX field will be their personal id. What goes in the VAT field?

**Comment by DocBenc** [comment]
Private customer has only TAX id with empty VAT id.
There are 2 main scenarios : private and company.
In the registration form the first question i put before all is "Buy as private [ ] or company [ ]"
If company is checked i ask all 4 needed fields to invoice i.e Company name, Vat id, Tax id and Uid code.
If private is checked the TAX id is not mandatory (unless he/she want invoice) as the seller does different kinds of tax documents. 

Concerning VAT enabled for Intra EU exchanges on Vies yes it is true, some EU VAT companies are not exchange enabled and in that case me italian cannot invoice in reverse charge to you so at the end you loose the VAT Tax in your country.

So the scenarious can be various depending also from local rules. I think that a world wide system could not be included natively in Zencart in a simple way. 

VAT id, TAX id should be included as they are used in many countries, the Uid code at the moment is for Italy and Portugal, soon Germany in 2021, France and Spain and other EU countries i think at the end of 2021 (limit is 31/12/2021).

**Comment by torvista** [comment]
I mis-understood the Uid was per invoice, but you are saying it is a single id the company has (from the tax authorities?) so it can use electronic invoicing, that has to appear on all invoices?

**Comment by DocBenc** [comment]
The Unique ID is assigned to each companies and work only for electronic invoice.

Try to explain.
.
The Unique ID identifies the endpoint handler of an electronic invoice. The handler is a container that may also be managed by a private accredited company. E.g many software house sell their ERP with credentials to SDI to manage electronic invoices. Each software house has an UniqueID of 7 characters. So the Unique ID is a sort of electronic address where the SDI delivers the invoices.
Different Companies may share the same Uid as the handling software is by the same software house. The Uid is obtained from the software you use to handle electronic invoicing.

Once the invoice is on destination handler it is sent to its final recipient using Taxid/Vatid pair.

You are not obliged to have a software because the tax agency provides you with the tools for invoices and their conservation free of charge in this case the Uid is 0000000.

However the Unique ID code is needed, default is 0000000. With default code invoice is delivered to your digital tax drawer (using tax id/vat id) where an italian citizen or company can access via smart card.

In case of a private buyer with only Tax Id the invoice is delivered to buyer digital tax drawer using default Uid.

As you understand all checks are done in advance by the SDI so every invoice must be compliant in its data or will be refused. A refused invoice is not existent by definition.and it's a serious trouble :-) .

**Comment by torvista** [comment]
Well, that's a wealth of valuable information you have provided. Thanks.

So, a tax id, a company name, a vat ID and a "vat id verified"  field would seem to be needed in ZC.
The Uid I feel is an aspect of the ERP/invoicing that is not necessarily part of ZC.

I note that currently the company name is in the address_book table and so not part of the customer definition.

Some thoughts from others, please, as how to proceed further on this?

**Comment by DocBenc** [comment]
You are welcome.
I use Zen since 15 years so i must only say thank you to you.

We italian are precursors of new EU e-invoice despite ourselves, so if you need further infos ask. I think that Uid will be widely used once e-invoice will be the standard as it allows to private software houses to be accredited on Exchange System and this will not be local. From what i know the Exchange System (SDI) used here has been developed in accordance with other Countries.

**Comment by torvista** [comment]
As this discussion would be better placed in the view of users rather than developers, I have opened a thread in the forum for further discussion:
https://www.zen-cart.com/showthread.php?226384-Addition-of-Customer-Company-Tax-VAT-fields&p=1365883#post1365883
